### DISCUSSION TITLE
How to leave all rooms the socket is currently in and join only one room?

### DISCUSSION THREAD

**Discussion by cedrickah** [socketio/socket.io#4905] [Q&A]
I'm using socket.io version 4.5.1, nodejs version 16

I'm trying to leave all rooms and subscribe to a new one, but I'm still receiving old events, I checked the content of io.of("/").adapter.rooms and when the socket join its added under the room but but when the socket leave the socket is not removed from the room.

**Comment by darrachequesne** [comment]
Hi! To leave all rooms:

```js
socket.rooms.forEach(room => {
  if (room !== socket.id) {
    socket.leave(room);
  }
});
```

Could you please provide some code reproducing the issue?

**Reply by cedrickah** [reply]
Of course

```
const rooms = io.of("/").adapter.rooms;
for (let room of rooms) {
    if (socket.id !== room[0]) {
      socket.leave(room[0]);
    }
}
```

It is say in the doc that rooms is now a Set...

**Reply by darrachequesne** [reply]
Looping over `io.of("/").adapter.rooms` should not be needed, the method [above](https://github.com/socketio/socket.io/discussions/4905#discussioncomment-8021896) should work as it only loops over the socket's current rooms. Could you please check?

**Reply by cedrickah** [reply]
I checked but none of these work. Neither meine nor yours. What I am trying to do really is to emit in only one room after I made sure the user is in only one room.
Here is the latest code I try but it emit to all sockets instead of the one in the room: 

```
io.on("connection", async (socket) => {
  
//leave all previous rooms before joining a new one
  socket.rooms.forEach((room) => {
    if (room !== socket.id) {
      socket.leave(room);
    }
  });
  
//join a new room
  socket.join(`${socket.handshake.query.exchangeId} room`);

  socket.on("new message", (data) => {
    io.to(data.room).emit("message response");
  });
});
```
